common/lib/core-interfaces/src/serializer.ts

```diff
@@ -28,12 +28,23 @@ export interface IFluidSerializer extends IProvideFluidSerializer {
      *
      * The original `input` object is not mutated.  This method will shallowly clones all objects in the path from
      * the root to any replaced handles.  (If no handles are found, returns the original object.)
+     *
+     * This will be deprecated and replaced in the future with encode.
      */
     replaceHandles(value: any, bind: IFluidHandle): any;
 
+    /**
+     * Given a mostly-plain object that may have handle objects embedded within, will return a fully-plain object
+     * where any embedded IFluidHandles have been replaced with a serializable form.
+     *
+     * The original `input` object is not mutated.  This method will shallowly clones all objects in the path from
+     * the root to any replaced handles.  (If no handles are found, returns the original object.)
+     */
+    encode?(value: any, bind: IFluidHandle): any;
```

Comment:
You can implement that method in the same PR just fine - classes can have extra methods :)
And you can validate (by linking repos and making it non-optional temporarily) that actually everything will work once versions are bumped and it becomes required argument.

Maybe it's right, but I hard hard time seeing next steps with optional method.
It feels like you want it to make non-optional, and force everyone to implement it on next bump of common-definitions, no?

Reply:
When I added decode, @anthony-murphy said it should be optional and not implemented where unnecessary (test objects that throw an error) to reduce the impact of the breaking change. I can update FluidSerializer at least where I do want to change it.

Reply:
I'm not sure I understand full sequence.
I assume the last step would be to have it non-optional and remove replaceHandles. What is the whole sequence?
I have no objection to it initially being optional, but I have a feeling that that this step does not give you much.

Here is what I see when I look at repo:
1.  ContainerRuntime.IFluidSerializer has been deprecated for ages and should be removed.
2. IFluidDataStoreRuntime.IFluidSerializer, while not marked deprecated, has been effectively deprecated and not used for ages
  - The only 3 places where it's used are experimental DDSs and mock - first should should be changed (see below), mock usage should be removed.
  - I do not see any usage in Bohemia repo either. Maybe worth double-checking Whiteboard as well to be sure before we remove it.
3. The only places where it is used is DDS layer, and there are no compat requirements here.

I.e. by fixing 1 & 2, we can completely remove IFluidSerializer and anything about it (including tests) from common and move them into runtime-utils or similar place and just fix all the usage / naming in one PR.

That's probably a bit more work, but I have a feeling that fixing # 2 is not that hard, and then it actually saves you time of not needing to bump various versions around.

Reply:
The remaining steps in the sequence would be (in subsequent released versions): use new methods if defined (otherwise using old), then make new methods required + mark old as deprecated + stop calling old, then remove deprecated. If there's no one using this or they are okay with a faster cadence of having to fix this (especially since this is an easy change), I agree that breaking it up like this doesn't help much.

Looking at Whiteboard codebase, they're only using IFluidSerializer in GenericSharedTree from SharedObject.

Your suggestion sounds good to me.

Reply:
Breaking any interface on IFluidObject, like IFluidSerializer creates a ton of pain due to #8077. This will cause problems for partners if we are not careful, as they frequently consume fluid from multiple places. Even outside IFluidObject breaking changes in any of our interfaces can cause problem for repos that consume fluid via multiple dependencies.  By being slow with breaking changes we remove the need for lockstep upgrade on the partner's side